jena-tdb/src/main/java/org/apache/jena/tdb/solver/SolverRX.java

```diff
@@ -18,298 +18,136 @@
-                return binding;
-            return null;
-        }
-
-        BindingNodeId b2 = new BindingNodeId(binding);
-        b2.put(var, tid);
-        return b2;
-    }
-
-    private static Node substitute(Node node, BindingNodeId binding, NodeTable nodeTable) {
-        if ( ! Var.isVar(node) )
-            return node;
-        Var var = Var.alloc(node);
-        try {
-            NodeId id = binding.get(var) ;
-            if ( id == null )
-                return null ;
-            if ( NodeId.isDoesNotExist(id) )
-                return null;
-            Node n = nodeTable.getNodeForNodeId(id) ;
-            if ( n == null )
-                // But there was to put it in the BindingNodeId.
-                throw new TDBException("No node in NodeTable for NodeId "+id);
-            return n ;
-        } catch (Exception ex)
-        {
-            FmtLog.error(SolverRX.class, ex, "SolverRX: substitute(%s) %s", node, binding) ;
-            return null ;
-        }
-    }
-
-    /**
-     * Match the NodeTupleTable with a tuple pattern.
-     * This is the accessor to the data.
-     * It assumes any triple terms have been dealt with.
-     */
-
-    private static Iterator<BindingNodeId> matchData(Iterator<BindingNodeId> chain, Tuple<Node> pattern, Args args) {
-        return SolverLib.solve(args.nodeTupleTable, pattern, args.anyGraph, chain, args.filter, args.execCxt);
-    }
-
-    private static Tuple<Node> createTuple(Tuple<Node> tuple, Var var, int idx) {
-        switch(idx) {
-            case 0: return TupleFactory.create3(var, tuple.get(1), tuple.get(2));
-            case 1: return TupleFactory.create4(tuple.get(0), var, tuple.get(2), tuple.get(3));
-            case 2: return TupleFactory.create3(tuple.get(0), tuple.get(1), var);
-            case 3: return TupleFactory.create4(tuple.get(0), tuple.get(1), tuple.get(2), var);
+    private static  Function<Tuple<NodeId>, Quad> asQuad(NodeTable nodeTable, int tupleLen, boolean anyGraph) {
+        switch (tupleLen) {
+            case 3:
+                return (Tuple<NodeId> t) -> {
+                    Node gx = Quad.defaultGraphIRI;
+                    Node sx = toNode(t.get(0), nodeTable);
+                    Node px = toNode(t.get(1), nodeTable);
+                    Node ox = toNode(t.get(2), nodeTable);
+                    return Quad.create(gx, sx, px, ox);
+                };
+            case 4:
+                return (Tuple<NodeId> t) -> {
+                    Node gx = (anyGraph)? Quad.unionGraph : toNode(t.get(0), nodeTable);
+                    Node sx = toNode(t.get(1), nodeTable);
+                    Node px = toNode(t.get(2), nodeTable);
+                    Node ox = toNode(t.get(3), nodeTable);
+                    return Quad.create(gx, sx, px, ox);
+                };
             default:
-                throw new TDBException("Index is not recognized: "+idx);
+                throw new InternalErrorException("Tuple of unknow length");
```

Comment:
s/unknow/unknown

Reply:
It's "uck" that there is code copy on but unifying TDB1 into DBOE/TDB2 framework risks destablizing the code.

Copying the code, so it is just import changes, is slightly nudging them together. But in nearby classes the code is appreciably different.

"Sometime".